### ISSUE TITLE
(torch-fourier-slice) Rationale behind multiplying by sinc2 grid in `project_3d_to_2d`

### ISSUE THREAD

**Issue by mgiammar** [teamtomo/teamtomo#65]
@alisterburt I remember you having some rationale (which I can't seem to remember) behind the sinc-squared grid multiplication during slice extraction:

https://github.com/teamtomo/torch-fourier-slice/blob/b53a76e0ae56e8086a36d929b6bd3ca4aeb149f1/src/torch_fourier_slice/project.py#L57-L65

I'm trying to re-update my mental model of how `torch-fourier-slice` is operating since there's some downstream computations whose numerical stability depends somewhat on having accurate projections. Would you mind re-explaining the rationale?

**Comment by alisterburt** [comment]
hey! oo this is a fun one, don't have much time right now so can't structure this explanation perfectly but

1. when you sample with linear interpolation the output you get is effectively convolved with the interpolation kernel which is a triangle function
2. the triangle function is the convolution of a rectangular pulse (e.g. single pixel response) with itself
3. convolution theorem: convolution in one domain is multiplication in the conjugate domain
4. sinc is the FT of rect
5. sinc2 is the FT of rect convolved with rect

we multiply/divide to 'undo' the effects of interpolation, it's imperfect but useful, I did some benchmarks with/without but don't have them handy

**Comment by alisterburt** [comment]
seems like this might be applied the wrong way though now that I'm thinking about it...

**Comment by mgiammar** [comment]
Makes sense from a signal processing perspective, I also see your point about it being applied the wrong way around. Interpolation in happening in Fourier space, so the anodization to correct for that fact should happen in real-space.

Looks like it is introducing some halo-ing for the no orientation offset case (first row).

<img width="889" height="1436" alt="Image" src="https://github.com/user-attachments/assets/65c07442-f048-4763-9720-a75c7b14e05a" />

**Comment by alisterburt** [comment]
could you compare to a straight projection of the map by summing along the z dimension instead of no sinc? That's a stronger control

Would also be good to see how pre division by sinc2 looks compared to pre multiplication

**Comment by alisterburt** [comment]
Might also be worth comparing directly to projection code in relion/warp

**Comment by mgiammar** [comment]
Line 296 of `src/projector.cpp` in RELION:

```
	// First do a gridding pre-correction on the real-space map:
	// Divide by the inverse Fourier transform of the interpolator in Fourier-space
	// 10feb11: at least in 2D case, this seems to be the wrong thing to do!!!
```

Then line 595 defines the function `Projector::griddingCorrect` which does the correction based on the radius of each voxel in real-space.

**Comment by mgiammar** [comment]
> could you compare to a straight projection of the map by summing along the z dimension instead of no sinc? That's a stronger control

There's a branch for direct 2D projection in ttsim3D which I tested with earlier today. But it's old, I don't remember all the decisions I made, and does not match precisely with the zero-orientation case. Can't be confident right now that I haven't made a mistake elsewhere in that code or this is a manifestation of the sinc2 difference.

**Comment by alisterburt** [comment]
Ah I just meant summing the simulated map (input for this library) along the depth dim e.g. torch.sum(volume, dim=-3)

**Comment by alisterburt** [comment]
The reason I'm suggesting that is we know that the frequency response  in that projection will be perfect in that direction, it will be a useful reference for comparison

Thanks for finding the relevant code in relion!

**Comment by mgiammar** [comment]
> Ah I just meant summing the simulated map (input for this library) along the depth dim e.g. torch.sum(volume, dim=-3)

Duh that is so much simpler sometimes you look past the obvious solution.

**Comment by mgiammar** [comment]
I've quickly tested out some of the projection corrections and have some of the data plotted below. The diagonals are projections from different methods, and the off-diagonals are the difference matrices.

Takeaways are that the direct summation case is within machine precision (makes sense, we hit the exact grid points in Fourier space for no orientation offset), but there are still ringing artifacts present in what is presumably the correct application of the sinc2 grid.

For the `sinc2_real_space` case, I've copied the implementation from RELION adding the following lines before doing the FFTs

```python
    # Calculate sinc^2 grid in real-space
    # First create normalized distance grid
    D = volume.shape[0]  # padded shape
    volume_shape = (D, D, D)
    midpoint = (D // 2, D // 2, D // 2)
    grid = coordinate_grid(
        image_shape=volume_shape,
        center=midpoint,
        norm=True,
        device=device,
    )
    grid = grid / D

    # Then, divide by sinc^2 where r is non-zero
    nonzero = grid > 0
    sinc = torch.sinc(grid)  # handles zero case internally
    volume = volume / (sinc**2)
```

<img width="898" height="865" alt="Image" src="https://github.com/user-attachments/assets/9a388c12-602e-4f38-8a4b-af9785e3932d" />

**Comment by michael-0brien** [comment]
Hello there! I'm the developer of [cryoJAX](https://github.com/michael-0brien/cryojax). Wonderful package (and ecosystem) you've all created here---we similarly are trying to develop extensible Python tools for cryo-EM :)

I happen to be also thinking about how to get sinc correction correct in Fourier slice extraction! I think that we also currently have a bug in cryoJAX for this.

I've based my implementation on cisTEM, where the relevant line is [here](https://github.com/timothygrant80/cisTEM/blob/master/src/core/image.cpp#L1691). Looking through the code and also the previous comment from @mgiammar, I believe there are still slight differences with the cisTEM version. Recapitulating cisTEM, I believe the implementation would be 

```python
D = volume.shape[0]
volume_shape = (D, D, D)
midpoint = (D // 2, D // 2, D // 2)
# Make *rectangular* grid in unit box
grid = coordinate_grid(
    image_shape=volume_shape,
    center=midpoint,
    norm=False, # opposed to `norm = True`
    device=device,
)
grid = grid / D
# For frequencies q = (q_x, q_y, q_x), Fourier transform of 3D rect is
# sinc(q_x) * sinc(q_y) * sinc(q_z), not sinc(|q|). Same should apply in real space?
grid_x, grid_y, grid_z = grid[..., 0], grid[..., 1], grid[..., 2]
sinc = torch.sinc(grid_x) * torch.sinc(grid_y) * torch.sinc(grid_z)
volume = volume / (sinc**2)
```

To be honest I haven't thought through what the inverse FT should be of the triangular function, and I'm not familiar with the RELION version, so I may be wrong! Though I'm fairly confident this is the relevant line in cisTEM (but it would be good to double check with them). Our current implementation is [here](https://github.com/michael-0brien/cryojax/blob/main/cryojax/ndimage/_transforms/_masks.py#L287), which is currently missing a square. It would be great to be in sync and make sure we also get this right over at cryoJAX.

FWIW I'll note cisTEM also implements the transformation to be average preserving. Their implementation seems to handle applying the transformation post application of a spherical mask, which I suspect is the reason for the average correction.

**Comment by alisterburt** [comment]
@michael-0brien great to see you here! Happy holidays, just wanted to say I'm super impressed by everything you're doing with cryojax :-)

I think it's pretty clear that the multiplication by sinc^2 here is a bug and dividing by sinc^2 in real space prior to dft calculation is closer to analytically correct

Useful note re: invariants - thank you - I guess ideally the signal scale is the same as you would get from a true projection and the spherical mask is to avoid artifacts from lack of periodicity in fourier space...

FWIW I've always seen this `project_3d_to_2d()` as a useful convenience layer but you often have to drop down to `extract_central_slices_rfft_3d()` for perf/specific control depending on the specific application, I don't think there is one right answer for how weighting should be handled 🙂

**Comment by michael-0brien** [comment]
> [@michael-0brien](https://github.com/michael-0brien) great to see you here! Happy holidays, just wanted to say I'm super impressed by everything you're doing with cryojax :-)

Much appreciated! Same goes for teamtomo, we rely on a few of these packages. `mmdf` was a lifesaver for us.

> I think it's pretty clear that the multiplication by sinc^2 here is a bug and dividing by sinc^2 in real space prior to dft calculation is closer to analytically correct

Yes---just note also the difference with cisTEM in computing $sinc(|r|)$ vs $sinc(r_x) * sinc(r_y) * sinc(r_z)$. If I'm not mistaken I think this difference is somewhat significant. This comes from the 3D fourier transform of the unit box.

> Useful note re: invariants - thank you - I guess ideally the signal scale is the same as you would get from a true projection and the spherical mask is to avoid artifacts from lack of periodicity in fourier space...

Yes, makes sense. Whether or not to do this seems quite dependent on the analysis pipeline!

> FWIW I've always seen this `project_3d_to_2d()` as a useful convenience layer but you often have to drop down to `extract_central_slices_rfft_3d()` for perf/specific control depending on the specific application, I don't think there is one right answer for how weighting should be handled 🙂

Agreed! We take a similar philosophy. It can be tricky to tie users to a fixed pattern for how diverse analysis pipelines can be. We also implement cubic interpolation which actually does a decent job handling the scaling issue, but this comes with its own significant downsides...

**Comment by mgiammar** [comment]
@michael-0brien I recently took a peek at the cryoJAX work and it's very exciting. Glad there are even more people out there interested in developing good open-source scientific software out there.

I have not had the brain-space recently to properly think about what prefilters are "correct" to apply, only that some projection results didn't match exactly what I expected and needed to know why. With the separability of  N-dim FFTs, I'd expect the the `sinc(r_x) * sinc(r_y) * sinc(r_z)` to be proper, but that _is not_ what the plots I've shared above were doing. I'll test this out when I get a chance.

**Comment by michael-0brien** [comment]
> [@michael-0brien](https://github.com/michael-0brien) I recently took a peek at the cryoJAX work and it's very exciting. Glad there are even more people out there interested in developing good open-source scientific software out there.

Thanks so much, back at you!

> I have not had the brain-space recently to properly think about what prefilters are "correct" to apply, only that some projection results didn't match exactly what I expected and needed to know why. With the separability of N-dim FFTs, I'd expect the the `sinc(r_x) * sinc(r_y) * sinc(r_z)` to be proper, but that _is not_ what the plots I've shared above were doing. I'll test this out when I get a chance.

Makes sense! Yes the separability is a good way to see it. Thanks so much for testing this out, I'm super interested to see what kinds of differences it makes.

**Comment by mgiammar** [comment]
Just had time to try out the sinc2 division across each of the dimensions separately which does produce realistic-looking projections but looking at the difference between a perfect projection (volume summed along dim=0), sinc2 division and no sinc2 division there are differences:

<img width="1196" height="688" alt="Image" src="https://github.com/user-attachments/assets/c3686640-ddca-4995-923c-0b8a9e1b3fb0" />

This case would have no interpolation artifacts since the interpolated points in Fourier space lie perfectly along the grid, so maybe not the best comparison. Still, though, this is an edge case the sinc2 grid should be able to handle.

**Comment by alisterburt** [comment]
@mgiammar fascinating 🧐 

Could you run a proj/brpoj cycle +- sinc2 and see how fsc to ground truth compares? Want to get a sense of how much it helps for arbitrary orientations
